File: Java/0101-0150/121. Best Time to Buy and Sell Stock.java
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/

class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        // Two Pointers
        int buy = 0, sell = 0;

        while(sell < prices.length){
            if(prices[buy] < prices[sell]){
                // update maxProfit
                maxProfit = Math.max(maxProfit, (prices[sell] - prices[buy]));
            }else{
                // update the buying price
                buy = sell;
            }
            sell++;
        }
        return maxProfit;
    }
}

File: Python/0101-0150/121. Best Time to Buy and Sell Stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy, sell, maxProfit = 0, 0, 0

        while sell < len(prices) :
            if prices[sell] > prices[buy]:
                # update the max profit
                maxProfit = max(maxProfit, (prices[sell] - prices[buy]))
            else:
                # update the buy
                buy = sell
            sell += 1
        
        return maxProfit
